[emby] Initial contribution

volfan6415 opened the pull request:
This request picks up on the initial contribution from https://github.com/openhab/openhab-addons/pull/7575 and rework that was done in https://github.com/openhab/openhab-addons/pull/13340.

I am finally back in a place where i am ready to get this over the finish line and incorporated.

This binding interfaces openHAB with a running instance of EMBY Media Server (https://emby.media/). EMBY is a PLEX alternative and the binding provides many of the same features for users of EMBY as the PLEX binding does for uses of PLEX.

I have used the versions of the binding which was previously submitted for over two years now (albeit in openHAB 2.X as the binding was not compatible until now with openHAB 3)(and as referenced in this thread [[New Binding] EMBY Server Binding - Alpha - Add-ons / Bindings - openHAB Community](https://community.openhab.org/t/new-binding-emby-server-binding-alpha/54010/94) I believe others in the community have been using it for awhile now as well. It has worked well for the users that have tested it for me. I have been using a version of this binding in my setup for the past 3 years now and it has been a great addition to my setup.

I have made the necessary updates to make the binding operate in openHAB 5 and now feel that this binding is ready to be incorporated. The core works well. While there are certainly features that can be added over time the binding in its current state more than covers the basics of what someone using openHAB would want to get from EMBY. It allows for interaction with the EMBY server to get currently playing devices/media which can then be used inside of openHAB as triggers for automations and the like.

I would like to flag that there will need to be clarification on is the author signing. In developing this binding I started with the KODI binding code, as EMBY provides a similar websocket api. While I did make extensive changes to work with the API provided by EMBY, there is definitively a structure left from the KODI binding and there are some methods and classes where the only real change was method/class names. Unfortunately in making these changes did not keep the author tags on the classes and now have all of the classes showing me as the original contribution. It is not my intention to not give credit where credit is due, and so will need some guidance on what edits need to be made to the author tags in the file to rectify this situation.

I am going to work on getting the readME in line as I await my review of this pull request.

Here is a link to test a version of the binding with openHAB 5

https://drive.google.com/file/d/1CSzcQHKiv1zfB5n1b6mSPXHCUbw2YMgt/view?usp=sharing

lsiepel commented:
Nice to see you continue to work on this contribution. Before we continue, please be sure that previous comments have been addressed. To speed up the review proces, i would like to ask you to perform a self-review by this checklist: https://github.com/openhab/openhab-addons/wiki/Review-Checklist

volfan6415 commented:
@lsiepel made the changes you suggested in the previous review.  I will do an analysis based on the code review guideline and post when I have completed that review and updated the reader.

volfan6415 commented:
@lsiepel

i have updated the readME and taken an initial pass at applying the coding review guidelines as you suggested.  I believe this is now ready for a review as the build checks are also now passing.

lsiepel commented on bundles/org.openhab.binding.emby/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,154 @@
+		<supported-bridge-type-refs>
+			<bridge-type-ref id="controller"/>
+		</supported-bridge-type-refs>
+		<label>EMBY Binding Device</label>
```
```suggestion
		<label>EMBY Device</label>
```

lsiepel commented on bundles/org.openhab.binding.emby/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,154 @@
+		<config-description>
+			<parameter name="deviceID" type="text" required="true">
+				<label>DeviceID</label>
+				<description>This is the deviceId you want to connect to to monitor.</description>
```
```suggestion
				<description>This is the deviceId you want to connect to.</description>
```

lsiepel commented on bundles/org.openhab.binding.emby/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,154 @@
+			<bridge-type-ref id="controller"/>
+		</supported-bridge-type-refs>
+		<label>EMBY Binding Device</label>
+		<description>This is a player device which connects to an EMBY server to play files that you want to monitor.</description>
```
```suggestion
		<description>This is a player device which connects to an EMBY server.</description>
```

lsiepel commented on bundles/pom.xml:
```diff
@@ -136,6 +131,7 @@
     <module>org.openhab.binding.electroluxappliance</module>
     <module>org.openhab.binding.elerotransmitterstick</module>
     <module>org.openhab.binding.elroconnects</module>
+    <module>org.openhab.binding.emby</module>
```
Only this change should be here, all other (remove white lines) should be reverted)

lsiepel commented on bundles/org.openhab.binding.emby/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,154 @@
+		</config-description>
+	</bridge-type>
+
+	<thing-type id="device">
```
Can you add equipment tags to both things?
For reference: https://next.openhab.org/docs/developer/bindings/thing-xml.html#thing-semantic-equipment-tags

volfan6415 replied:
done

lsiepel reviewed (changes requested):
Thansk for your contribution. Covered not all files, 5 left. Will finish the review once these are fixed.

volfan6415 commented on bundles/org.openhab.binding.emby/NOTICE:
```diff
@@ -0,0 +1,12 @@
+terms of the Eclipse Public License 2.0 which is available at:
+https://www.eclipse.org/legal/epl-2.0/
+
+Dependencies:
+- org.lastnpe.eea:eea-all:2.4.0
+- org.apache.karaf.features:framework:4.4.7
```
fixed

volfan6415 replied:
fixed

lsiepel replied:
The contents are not the default contents it should have, please check or copy from other binding like plugwiseha

volfan6415 replied:
done copied from plugwiseha

volfan6415 commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,188 @@
+- **`emby:controller`**  
+  Represents a connection to an Emby server (a Bridge Thing).
+
+- **`emby:device`**  
```
fixed

volfan6415 commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,188 @@
+
+# Discovery
+
+The binding supports automatic discovery for both servers and clients.
+
+## Bridge Discovery
+
+Emby servers are automatically detected on the local network.
+
+They are offered as **`emby:controller`** Things.
+
+## Client Discovery
+
+When an **`emby:controller`** Bridge is online, connected Emby clients are automatically discovered.
+
+They are offered as **`emby:device`** Things under the respective bridge.
```
fixed

lsiepel replied:
The paragraph header was also deleted, that should be reverted.

volfan6415 replied:
re added paragraph header Automatic Discovery

volfan6415 commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,188 @@
+
+The following Configuration Parameter Keys are available:
+
+| Name | Type | Description | Default | Required | Advanced |
```
fixed

volfan6415 commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,188 @@
+
+The following Channel IDs are available for an `emby:device` Thing:
+
+| Channel ID | Item Type | Config Parameters | Description |
```
changed

lsiepel replied:
Improved, but not fully aligned yet,

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/EmbyBridgeConfiguration.java:
```diff
@@ -0,0 +1,39 @@
+public class EmbyBridgeConfiguration extends Configuration {
+
+    public String api = "";
+    public int bufferSize = 10000; // Default websocket buffer size
+    public String ipAddress = "";
+    public int port = 8096; // Default server port
+    public int refreshInterval = 10000; // Default refresh interval
+    public boolean discovery = true; // Discovery enabled by default
```
they are now documented in the readme

lsiepel replied:
port is missing from the documentation

volfan6415 replied:
added port

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/EmbyHandlerFactory.java:
```diff
@@ -0,0 +1,148 @@
+        callbackUrl = (String) properties.get("callbackUrl");
+    }
+
+    @SuppressWarnings("null")
```
removed

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/discovery/EmbyBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,190 @@
+                logger.debug("Discovery interrupted, exiting");
+            }
+
+            // Broadcast message over all network interfaces
```
updated to utilize these methods

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyActions.java:
```diff
@@ -0,0 +1,164 @@
+    @RuleAction(label = "@text/action.emby.SendPlay.label", description = "@text/action.emby.SendPlay.desc")
+    public void sendPlay(
+            @ActionInput(name = "itemIds", label = "@text/action.emby.SendPlay.input.itemIds.label", description = "@text/action.emby.SendPlay.input.itemIds.desc") String itemIds,
+
```
done

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/protocol/EmbyHttpRetryExceeded.java:
```diff
@@ -0,0 +1,29 @@
+ *
+ * @author Zachary Christiansen - Initial contribution
+ */
+@SuppressWarnings("serial")
```
fixed

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -0,0 +1,13 @@
+		The Emby Binding integrates Emby, a personal media server (https://emby.media/), with openHAB. It supports
+		both controlling the player and retrieving player status data like the currently playing movie title.
+	</description>
+
```
done

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/EmbyBridgeListener.java:
```diff
@@ -0,0 +1,33 @@
+@NonNullByDefault
+public interface EmbyBridgeListener extends EventListener {
+
+    void updateConnectionState(boolean connected);
```
done

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/EmbyBridgeListener.java:
```diff
@@ -0,0 +1,33 @@
+ * @author Zachary Christiansen - Initial Contribution
+ */
+@NonNullByDefault
+public interface EmbyBridgeListener extends EventListener {
```
The problem is that I already have an EmbyConnection.java which implements the embyclientsocketevent listner.   

This really is named right as this is what the bridge implements

lsiepel replied:
hmm ok. i can understand the confusion.

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/EmbyEventListener.java:
```diff
@@ -0,0 +1,68 @@
+        CLEAR
+    }
+
+    void updateConnectionState(boolean connected);
```
done

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/protocol/EmbyClientSocketEventListener.java:
```diff
@@ -0,0 +1,32 @@
+@NonNullByDefault
+public interface EmbyClientSocketEventListener {
+
+    void handleEvent(EmbyPlayStateModel playstate);
```
done

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/protocol/EmbyDeviceEncoder.java:
```diff
@@ -0,0 +1,30 @@
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
+/**
+ * EmbyClientSocket implements the low level communication to Emby through
+ * websocket. Usually this communication is done through port 9090
```
fixed

volfan6415 commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,188 @@
+| **`showtitle`** | String | None | Title of the currently playing movie or TV show. |
+| **`mute`** | Switch | None | Mute status control. |
+| **`imageurl`** | String | `imageurl_maxHeight`, `imageurl_maxWidth`, `imageurl_type` | URL for current media artwork. |
+| **`currenttime`** | Number:Time | None | Current playback position. |
```
done

lsiepel replied:
Could you remove the markup for the channel id's
**`control`** => control

volfan6415 commented:
@lsiepel i have made all of the changes that you have suggested.  I don't know why the build is failing on the runner with a spotless check.  I have run that on the version on my system and it passes and then the build also passes.

lsiepel commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,163 @@
+# Emby Binding
+
+The **Emby Binding** integrates [Emby](https://emby.media/), a personal media server, with openHAB.
+
+It allows controlling Emby players and retrieving player status data.
+
+For example, you can monitor the currently playing movie title or automatically dim your lights when playback starts.
+
+This binding supports multiple Emby clients connected to a single Emby Media Server.
+
+It provides functionality similar to the Plex Binding.
+
+---
```
Remove whitelines and dashes
```suggestion
It allows controlling Emby players and retrieving player status data.
For example, you can monitor the currently playing movie title or automatically dim your lights when playback starts.
This binding supports multiple Emby clients connected to a single Emby Media Server.
It provides functionality similar to the Plex Binding.
```

volfan6415 replied:
done

lsiepel commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,163 @@
+
+This binding defines the following Thing Type IDs:
+
+- **`controller`**  
+  Represents a connection to an Emby server (a Bridge Thing).
+
+- **`device`**  
+  Represents a client/player device connected to the Emby server.
+
+---
```
```suggestion
- `controller`
  Represents a connection to an Emby server (a Bridge Thing).

- `device`
  Represents a client/player device connected to the Emby server.
```

volfan6415 replied:
removed markup

lsiepel commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,163 @@
+
+| Name                  | Type    | Description                                             | Default | Required | Advanced |
+|-----------------------|---------|---------------------------------------------------------|---------|----------|----------|
+| **`ipAddress`**       | Text    | IP address or hostname of the Emby server.              | N/A     | Yes      | No       |
```
Remove markup from parameter id
**`ipAddress`** => ipAddress

lsiepel commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,163 @@
+| **`current-time`** | Number:Time | None                                                       | Current playback position.                                            |
+| **`duration`**    | Number:Time | None                                                       | Total media duration.                                                 |
+| **`media-type`**   | String      | None                                                       | Type of media (e.g., Movie, Episode).                                 |
+
+---
```
```suggestion
```

volfan6415 replied:
removed

lsiepel commented on bundles/org.openhab.binding.emby/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,156 @@
+	<bridge-type id="controller">
+		<label>EMBY Server</label>
+		<description>This is the Bridge to an instance of an EMBY server you want to connect to.</description>
+		<semantic-equipment-tag>Computer</semantic-equipment-tag>
```
I think `NetworkAppliance` is a bettert fit

volfan6415 replied:
done

lsiepel commented on only_nonws.patch:
Remove this file.

volfan6415 replied:
removed

lsiepel commented on bundles/pom.xml:
```diff
@@ -768,4 +769,4 @@
     </profile>
   </profiles>
 
-</project>
+</project>
```
Revert this, as this is causing the build error. it is outside of the bundle directory and therefore you don't experience this locally when building from the project subfolder.

volfan6415 replied:
done

volfan6415 commented:
changes made i keep fighting the bundles/pom.xml.  If it only has the one suggested change to add my binding the build check here fails.  Where as if i run mvn spotless apply at the root it adds a space at the end but then the online build passess.  Open to suggestions.

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/EmbyHandlerFactory.java:
```diff
@@ -0,0 +1,147 @@
+            .unmodifiableSet(Stream.of(THING_TYPE_EMBY_CONTROLLER, THING_TYPE_EMBY_DEVICE).collect(Collectors.toSet()));
+    private @Nullable String callbackUrl = null;
+    private final Map<ThingUID, @Nullable ServiceRegistration<?>> discoveryServiceRegs = new HashMap<>();
+
```
With this change, you can cleanup:
- remove the `@NonNullByDefault({})` annotations to the two private vars.
- remvoe the set and unset overriden methods. 


```suggestion

    @Activate
    public EmbyHandlerFactory(final @Reference WebSocketFactory webSocketFactory,
            final @Reference NetworkAddressService networkAddressService) {
        this.webSocketFactory = webSocketFactory;
        this.typeGenerator = typeGenerator;
    }
```

volfan6415 replied:
done

lsiepel commented on bundles/org.openhab.binding.emby/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,156 @@
+		<description>The url of the playing media</description>
+		<state readOnly="true" pattern="%s"/>
+		<config-description>
+			<parameter name="image-url_type" type="text">
```
These parameters need some work, this comment is just for reference to the other comment in the binding constants.

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/EmbyBindingConstants.java:
```diff
@@ -0,0 +1,74 @@
+
+    public static final String CHANNEL_DURATION = "duration";
+    public static final String CHANNEL_IMAGEURL = "image-url";
+    public static final String CHANNEL_IMAGEURL_TYPE = "image-url_type";
+    public static final String CHANNEL_IMAGEURL_MAXWIDTH = "image-url_maxWidth";
+    public static final String CHANNEL_IMAGEURL_MAXHEIGHT = "image-url_maxHeight";
+    public static final String CHANNEL_IMAGEURL_PERCENTPLAYED = "image-url_PercentPlayed";
```
These seem like configuration parameters if i look at the thing-type.xml
1. parameters should have `camelCase` convention.
   For reference: https://www.openhab.org/docs/developer/guidelines.html#naming-convention
2. These parameters are not documented
3. How does this work ? i would not expect PercentPlayed as a configuration parameter. 
4. PercentPlayed is not listed in the thing-type.xml

volfan6415 replied:
They are configuration parameters.   I will make the changes to add CONFIG in front of them and also add to the documentation

As to PERCENTPLAYED.  I will add to the readme to explain the function of this.  But it is a boolean parameter that allows the image url that is retrieved to have the following added onto the end 

Adding a Percent Played Indicator
A percent played overlay can be added onto the image using PercentPlayed=47

This changes the image that is retrieved.

volfan6415 replied:
done

lsiepel commented on bundles/pom.xml:
```diff
@@ -767,5 +768,4 @@
       </build>
     </profile>
   </profiles>
-
```
You can copy paste from main branch? this line should not be removed.

lsiepel reviewed (changes requested):
Soem additional/follow up comments to the previous review round.

volfan6415 commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,160 @@
+| bufferSize      | Integer | WebSocket buffer size in bytes.                         | 10,000  | No       | No       |
+| refreshInterval | Integer | Polling interval for play-state updates (milliseconds). | 10,000  | No       | No       |
+| port            | Integer | Port in which EMBY is listening for communication.      | 8096    | No       | No       |
+
+| **`discovery`** | Boolean | Enable or disable automatic device discovery. | true | No | Yes |
```
done

lsiepel replied:
Di you push your change? I don't see it.

volfan6415 commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,160 @@
+
+The following Configuration Parameter Key is available:
+
+- **`deviceID`**  
```
done

volfan6415 commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,160 @@
+
+# Channels
+
+The following Channel IDs are available for an `emby:device` Thing:
```
done

lsiepel replied:
Partial fixed. an => a

volfan6415 replied:
made the suggested chance

volfan6415 commented on bundles/org.openhab.binding.emby/README.md:
````diff
@@ -0,0 +1,160 @@
+}
+```
+
+## Item Configuration Example
````
done

volfan6415 commented on bundles/org.openhab.binding.emby/README.md:
````diff
@@ -0,0 +1,160 @@
+String      Emby_ImageURL    "Artwork URL [%s]"  { channel="emby:device:myEmbyServer:myClientDevice:image-url" }
+```
+
+## Sitemap Configuration Example
````
done

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/EmbyBindingConstants.java:
```diff
@@ -0,0 +1,74 @@
+
+    // Module Properties
+
+    public static final String HOST_PARAMETER = "ipAddress";
+    public static final String WS_PORT_PARAMETER = "port";
+    public static final String WS_BUFFER_SIZE = "bufferSize";
+    public static final String HTTP_PORT_PARAMETER = "httpPort";
+    public static final String HTTP_USER_PARAMETER = "httpUser";
+    public static final String HTTP_PASSWORD_PARAMETER = "httpPassword";
+    public static final String REFRESH_PARAMETER = "refreshInterval";
+    public static final String API_KEY = "api";
+    public static final String DEVICE_ID = "deviceID";
+    public static final String DISCOVERY_ENABLE = "discovery";
```
made the change and removed the parameters that were not used

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyBridgeHandler.java:
```diff
@@ -0,0 +1,307 @@
+    @Reference
+    private @Nullable TranslationProvider i18nProvider;
+
+    public EmbyBridgeHandler(Bridge bridge, @Nullable String hostAddress, @Nullable String port,
```
```suggestion
    public EmbyBridgeHandler(Bridge bridge, @Nullable String callbackAddress, @Nullable String callbackPort,
```

volfan6415 replied:
revised

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyBridgeHandler.java:
```diff
@@ -0,0 +1,307 @@
+            }
+            this.connection = new EmbyConnection(this, this.webSocketClient);
+            try {
+                EmbyBridgeConfiguration cfg = checkConfiguration();
```
I would prefer to perform a configuration validation at the start of this method. after those checks, set the thingstatus to unknown and start the async task to establish a connection.

volfan6415 replied:
reordered and added thing status updates

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyBridgeHandler.java:
```diff
@@ -0,0 +1,307 @@
+        } else {
+            reconnectionCount++;
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "The connection to the emby server was closed, binding will wait one min and attempt to restablish connection. There have been "
+                            + Integer.toString(reconnectionCount) + " attempts to establish a new connection");
```
Shorter detailed state is better. Maybe something like:
Connection lost. Retrying in 1 minute. Attempt x

volfan6415 replied:
revised

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyBridgeHandler.java:
```diff
@@ -0,0 +1,307 @@
+
+        if (embyConfig.api == null || embyConfig.api.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "There is no API key configured for this bridge, please add an API key to enable communication");
```
1. Shorter is usually better. The user knowns the thing is offline due to a configuraiton error. The user only needs to know what parameter. 
2. Did you consider adding i18n custom keys to all thing state updates? It would enable translation of these detailed states. for reference: https://www.openhab.org/docs/developer/utils/i18n.html#using-custom-keys

volfan6415 replied:
revised all thing status messages and updated to use i18n custom keys

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyDeviceHandler.java:
```diff
@@ -0,0 +1,562 @@
+ */
+package org.openhab.binding.emby.internal.handler;
+
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CHANNEL_CONTROL;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CHANNEL_CURRENTTIME;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CHANNEL_DURATION;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CHANNEL_IMAGEURL;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CHANNEL_IMAGEURL_MAXHEIGHT;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CHANNEL_IMAGEURL_MAXWIDTH;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CHANNEL_IMAGEURL_PERCENTPLAYED;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CHANNEL_IMAGEURL_TYPE;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CHANNEL_MEDIATYPE;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CHANNEL_MUTE;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CHANNEL_SHOWTITLE;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CHANNEL_STOP;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CHANNEL_TITLE;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CONTROL_GENERALCOMMAND;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CONTROL_MUTE;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CONTROL_PAUSE;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CONTROL_PLAY;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CONTROL_SENDPLAY;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CONTROL_SESSION;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CONTROL_STOP;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.CONTROL_UNMUTE;
+import static org.openhab.binding.emby.internal.EmbyBindingConstants.DEVICE_ID;
```
wildcards are allowed for static imports
```suggestion
import static org.openhab.binding.emby.internal.EmbyBindingConstants.*;
```

volfan6415 replied:
revised

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyDeviceHandler.java:
```diff
@@ -0,0 +1,562 @@
+    private @Nullable EmbyDeviceConfiguration config;
+    private @Nullable EmbyPlayStateModel currentPlayState = null;
+    private @Nullable EmbyBridgeHandler bridgeHandler;
+    private static final long MIN_UPDATE_INTERVAL_MS = 1000; // 1 second
```
drop or use

volfan6415 replied:
revised

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyDeviceHandler.java:
```diff
@@ -0,0 +1,562 @@
+                    return UnDefType.UNDEF;
+                }
+                Boolean muted = currentPlayState.getEmbyMuteSate();
+                return Boolean.TRUE.equals(muted) ? OnOffType.ON : OnOffType.OFF;
```
```suggestion
                return OnOffType.from(currentPlayState.getEmbyMuteSate().booleanValue());
```

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyDeviceHandler.java:
```diff
@@ -0,0 +1,562 @@
+
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.NONE, "Initializing Emby device");
```
```suggestion
        updateStatus(ThingStatus.UNKNOWN);
```

volfan6415 replied:
revised

lsiepel reviewed (changes requested):
Covered the remaining files. There are also many build warnings, and they should all be fixed. Most null pointer 
 warnings can probably be fixed by using a local variable.

lsiepel reviewed (approved):
Thanks. LGTM.
As some vital connection handling has been reworked. Do you need additional test time before a merge?

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/protocol/EmbyConnection.java:
```diff
@@ -0,0 +1,145 @@
+     */
+    public synchronized void connect(String httpBaseUrl, String apiKey, ScheduledExecutorService sched, int refreshRate,
+            int bufferSize) {
+
```
done

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/protocol/EmbyConnection.java:
```diff
@@ -0,0 +1,145 @@
+        }
+    }
+
+    public synchronized void close() {
```
added

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyBridgeHandler.java:
```diff
@@ -0,0 +1,307 @@
+    private @Nullable EmbyBridgeConfiguration config;
+    private int reconnectionCount;
+    private @Nullable String lastDiscoveryStatus = null;
+    private int port;
```
corrected

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyDeviceHandler.java:
```diff
@@ -0,0 +1,562 @@
+        updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.NONE, "Initializing Emby device");
+        scheduler.execute(() -> {
+            try {
+                EmbyDeviceConfiguration cfg = validateConfiguration();
+                this.config = cfg;
```
revised the inialize method so that now i don't need to set a local variable of the cfg

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyDeviceHandler.java:
```diff
@@ -0,0 +1,562 @@
+                    return;
+                }
+
+                this.bridgeHandler = (EmbyBridgeHandler) bridge.getHandler();
```
removed

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyDeviceHandler.java:
```diff
@@ -0,0 +1,562 @@
+
+                    boolean newMuteState = playstate.getEmbyMuteSate();
+                    if (newMuteState != lastMuted) {
+                        updateMuted(newMuteState);
```
revised

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyDeviceHandler.java:
```diff
@@ -0,0 +1,562 @@
+    }
+
+    private State createStringState(@Nullable String string) {
+        if (string == null || string.isEmpty()) {
```
revised

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyBridgeHandler.java:
```diff
@@ -0,0 +1,307 @@
+                                "Connection could not be established");
+                    }
+                }, CONNECTION_CHECK_INTERVAL_MS, CONNECTION_CHECK_INTERVAL_MS, TimeUnit.MILLISECONDS);
+
```
removed

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyBridgeHandler.java:
```diff
@@ -0,0 +1,307 @@
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "There is no API key configured for this bridge, please add an API key to enable communication");
+            throwValidationError("api", "Missing value for: api");
+
```
revised

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyBridgeHandler.java:
```diff
@@ -0,0 +1,307 @@
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "No network address specified, please specify the network address for the EMBY server");
+            throwValidationError("ipAddress", "Missing value for: ipAddress");
+
```
revised

volfan6415 commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/protocol/EmbyClientSocket.java:
```diff
@@ -0,0 +1,233 @@
+            session = null;
+        }
+        try {
+            client.stop();
```
removed the stop and start calls

volfan6415 commented:
> Covered the remaining files. There are also many build warnings, and they should all be fixed. Most null pointer warnings can probably be fixed by using a local variable.

fixed the build warnings

volfan6415 commented:
> Thanks. LGTM. As some vital connection handling has been reworked. Do you need additional test time before a merge?

No i have been testing the binding as the changes were being made and the binding is working as intended with openhab 5.0 M2.

volfan6415 commented:
Ok i have addressed the review comments and all of the warnings during the build.  No further testing is needed as I have confirmed that the binding is operating as intended.

@lsiepel the last two commits on this branch indicate the issues that I am having with the pom.xml

@volfan6415
[Merge branch 'openhab:main' into Emby-Initial-Binding](https://github.com/openhab/openhab-addons/pull/18607/commits/72ae06adc11cfdd43fb5b6943178919722c0cdb2)
72ae06a
@volfan6415
[Update pom.xml](https://github.com/openhab/openhab-addons/pull/18607/commits/7bdeeb89a0d5544753320ac0abe81dc0df105573) 
7bdeeb8

As you can see the first commit were the pom has the line removed.  The build succeeds.  The most recent commit has only the one line changed but the build fails.

What is the best path forward to fix this.

volfan6415 commented:
@lsiepel i spoke too soon.  I am getting some weird behavior where the connection reconnection race condition gets started and the event logger gets overwhelmed.  I have pushed up a possible fix for this issue but want to test it for a few days and get to the bottom of the issue.

here is a snippet of my logs when the problem happened

2025-05-03 20:27:27.078 [INFO ] [openhab.event.ThingStatusInfoEvent  ] - Thing 'emby:device:[DEVICE-ID]:[DEVICE-SUFFIX-A]' updated: OFFLINE (BRIDGE_OFFLINE)
2025-05-03 20:27:27.078 [WARN ] [ab.core.internal.events.EventHandler] - EventLogger queue exceeds 5000 elements. System may be unstable.
2025-05-03 20:27:27.078 [INFO ] [openhab.event.ThingStatusInfoEvent  ] - Thing 'emby:device:[DEVICE-ID]:[DEVICE-SUFFIX-B]' updated: OFFLINE (BRIDGE_OFFLINE)
2025-05-03 20:27:27.078 [INFO ] [openhab.event.ThingStatusInfoEvent  ] - Thing 'emby:controller:[DEVICE-ID]' updated: OFFLINE (COMMUNICATION_ERROR): Connection closed (retry: 13099)
2025-05-03 20:27:27.079 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Controller state changed: Connection closed (retry 13097 → 13098)

... (⚠️ 92 repeated log lines removed for brevity) ...

2025-05-03 20:27:27.085 [INFO ] [openhab.event.ThingStatusInfoEvent  ] - Thing 'emby:controller:[DEVICE-ID]' updated: OFFLINE (COMMUNICATION_ERROR): Connection closed (retry: 13108)
2025-05-03 20:27:27.087 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Controller state changed: Connection closed (retry 13108 → 13109)

lsiepel commented:
Build fails due to spotless:
```
Failed to execute goal com.diffplug.spotless:spotless-maven-plugin:2.44.3:check (codestyle_check) on project org.openhab.binding.emby: Unable to check file /home/runner/work/openhab-addons/openhab-addons/bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyBridgeHandler.java: com.google.googlejavaformat.java.FormatterException: 186:97: error: ';' expected -> [Help 1]
```

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/EmbyHandlerFactory.java:
```diff
@@ -0,0 +1,132 @@
+
+    @Activate
+    public EmbyHandlerFactory(@Reference WebSocketFactory webSocketClientFactory,
+            @Reference NetworkAddressService networkAddressService, ComponentContext componentContext) {
```
This is no longer used, so can be cleaned-up
```suggestion
            ComponentContext componentContext) {
```

volfan6415 replied:
removed

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/protocol/EmbyConnection.java:
```diff
@@ -0,0 +1,163 @@
+        }
+
+        listener.updateConnectionState(false);
+        logger.info("Disposed EmbyConnection for {}:{}", hostname, embyport);
```
```suggestion
        logger.debug("Disposed EmbyConnection for {}:{}", hostname, embyport);
```

volfan6415 replied:
changed

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/protocol/EmbyClientSocket.java:
```diff
@@ -0,0 +1,209 @@
+        }
+        reconnectAttempts++;
+        long delay = Math.min(60_000, (1 << Math.min(reconnectAttempts, 6)) * 1000L);
+        logger.info("Scheduling reconnect #{} in {}ms", reconnectAttempts, delay);
```
info level should be used as less as possible. only iof really import to the user
```suggestion
        logger.debug("Scheduling reconnect #{} in {}ms", reconnectAttempts, delay);
```

volfan6415 replied:
changed

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/protocol/EmbyClientSocket.java:
```diff
@@ -0,0 +1,209 @@
+            logger.debug("already open");
+            return;
+        }
+        if (!client.isStarted()) {
+            client.setMaxTextMessageBufferSize(bufferSize);
+        }
```
1. The client will always be started here.
2. I don't think you should set this as this is globally for all bindings.

if this is really needed, you should not use the shared client.

volfan6415 replied:
removed it.  Which also meant that config parameters for setting the buffer size are no longer needed so those have been removed as well.  

i think this was a hold over from my early testing with developing the binding where i thought that the buffer size needed to be larger when in fact it does not need to be.

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/protocol/EmbyClientSocket.java:
```diff
@@ -0,0 +1,209 @@
+                "Future<Session>.get returned null Session");
+        requireNonNull(wsSession, "WebSocketClient.connect returned null Session");
+        this.session = wsSession;
+        logger.info("Connected to Emby");
```
```suggestion
        logger.debug("Connected to Emby");
```

volfan6415 replied:
changed

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyBridgeHandler.java:
```diff
@@ -0,0 +1,277 @@
+    private @Nullable EmbyBridgeConfiguration config;
+    private int reconnectionCount;
+    private @Nullable String lastDiscoveryStatus;
+    @Reference
+    private @NonNullByDefault({}) TranslationProvider i18nProvider;
```
Please inject this in the constructor / factory similar to the `WebSocketFactory`

volfan6415 replied:
changed for both the bridge handler and the device handler

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyBridgeHandler.java:
```diff
@@ -0,0 +1,277 @@
+                    }
+                }, CONNECTION_CHECK_INTERVAL_MS, CONNECTION_CHECK_INTERVAL_MS, TimeUnit.MILLISECONDS);
+            } catch (Exception e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
```
multiple configuration issues are now routed to COMMUNICATION_ERROR. Please make sure the configuration errors end up as CONFIGURATION_ERROR thingstates.

volfan6415 replied:
they do.  establishConnection() does not get called until after the validate configuration has passed.  So if there are any errors in that they get caught as config validation errors and the loop is stopped and the proper thing messages are set .  See 

    @Override
    public void initialize() {
        updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.NOT_YET_READY);

        final ScheduledExecutorService exec = requireNonNull(scheduler, "scheduler must not be null");

        exec.execute(() -> {
            this.reconnectionCount = 0;
            try {
                this.config = checkConfiguration();
                this.connection = new EmbyConnection(this, this.webSocketClient);
                establishConnection();
                updateStatus(ThingStatus.ONLINE);
            } catch (ConfigValidationException cve) {
                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
                        "@text/thing.status.bridge.configurationFailed" + cve.getMessage());
            }
        });
    }
    
    so the flow is we validate the config ... if it passes then we try and establish a connection ... thus the exception that is referenced above is truly just to catch communication errors.

lsiepel commented on bundles/org.openhab.binding.emby/src/main/java/org/openhab/binding/emby/internal/handler/EmbyDeviceHandler.java:
```diff
@@ -0,0 +1,542 @@
+                }
+
+                updateStatus(ThingStatus.ONLINE);
+
```
```suggestion
```

volfan6415 replied:
removed

lsiepel reviewed (changes requested):
The new commits need some more work.

volfan6415 commented:
ok made the edits suggested above.  And the build errors on my local machine have gone away.

so far the binding in its current configuration in my testing has been stable and is acting as intended.  I have not seen the race condition and i think that the web socket connection methods have now stabilized.

I want to test for another day or two to confirm and I will post back once that is confirmed.

volfan6415 commented:
also the bundles/pom.xml issue appears to have finally gone away.

volfan6415 commented:
@lsiepel ok i have used the binding has been stable for three days straight.  I have also simulated emby server disconnects (shutdown the connected server) and the binding didn't create a race condition and reconnected gracefully when the emby server started back up.

from a testing perspective i would say this binding is good to go.

lsiepel commented:
Two comments are still open:
https://github.com/openhab/openhab-addons/pull/18607/files#r2071880949
https://github.com/openhab/openhab-addons/pull/18607#discussion_r2072540558

Otherwise LGTM!

volfan6415 commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,156 @@
+This binding supports multiple Emby clients connected to a single Emby Media Server.
+It provides functionality similar to the Plex Binding.
+
+# Supported Things
```
made the suggested change to this and the other headings

lsiepel replied:
not sure what you did, but the suggested change was not added. (onely one top level paragraph is allowed)
```suggestion
## Supported Things
```

volfan6415 commented:
@lsiepel i belive i have made the changes requested

lsiepel reviewed (approved):
Thansk, LGTM.

Now, you could add the binding's logo to the openHAB website. See https://www.openhab.org/docs/developer/addons/#add-your-add-on-s-logo-to-the-openhab-website

lsiepel commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,164 @@
+- `deviceID`  
+  The unique identifier for the client device connected to the Emby server.
+
+# Channels
```
```suggestion
## Channels
```

lsiepel commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,164 @@
+
+# Full Example
+
+## `emby.things` Example
```
```suggestion
### `emby.things` Example
```

lsiepel commented on bundles/org.openhab.binding.emby/README.md:
````diff
@@ -0,0 +1,164 @@
+String      Emby_ImageURL    "Artwork URL [%s]"  { channel="emby:device:myEmbyServer:myClientDevice:image-url" }
+```
+
+## `emby.sitemap` Configuration Example
````
```suggestion
### `emby.sitemap` Configuration Example
```

lsiepel commented on bundles/org.openhab.binding.emby/README.md:
````diff
@@ -0,0 +1,164 @@
+}
+```
+
+# Rule Actions
````
```suggestion
## Rule Actions
```

lsiepel commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,164 @@
+
+All playback and control commands are now implemented as Rule Actions rather than channels. Use the standard `getActions` API in your rules to invoke these.
+
+## Available Actions
```
```suggestion
### Available Actions
```

lsiepel commented on bundles/org.openhab.binding.emby/README.md:
```diff
@@ -0,0 +1,164 @@
+| sendGeneralCommand         | `sendGeneralCommand(CommandName: String)`                                                                                                                                               | Send a generic Emby control command (e.g., MoveUp, ToggleMute, GoHome).                                 |
+| sendGeneralCommandWithArgs | `sendGeneralCommandWithArgs(CommandName: String, Arguments: String)`                                                                                                                    | Send a generic Emby control command with a JSON arguments blob (e.g., SetVolume, DisplayMessage, etc.). |
+
+# Example Rule (XTend)
```
```suggestion
### Example Rule (XTend)
```

lsiepel commented on bundles/org.openhab.binding.emby/README.md:
````diff
@@ -0,0 +1,164 @@
+end
+```
+
+# Example Rule (JavaScript)
````
```suggestion
### Example Rule (JavaScript)
```

lsiepel commented on bundles/org.openhab.binding.emby/README.md:
````diff
@@ -0,0 +1,164 @@
+emby.sendGeneralCommand("ToggleMute");
+```
+
+# References
````
```suggestion
## References
```

volfan6415 commented:
@lsiepel i have committed the paragraph suggestions i think I have now addressed everything

lsiepel reviewed (approved):
Thanks, LGTM.

Now, you could add the binding's logo to the openHAB website. See https://www.openhab.org/docs/developer/addons/#add-your-add-on-s-logo-to-the-openhab-website
